correct mapping for postgres timestamptz type to sql type TIMESTAMP_W…

lopata2 opened the pull request:
…ITH_TIMEZONE and not to TIMESTAMP

### All Submissions:

* [ ] Have you followed the guidelines in our [Contributing](https://github.com/pgjdbc/pgjdbc/blob/master/CONTRIBUTING.md) document?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

PR is solving this the issue described here https://github.com/pgjdbc/pgjdbc/issues/1766 .

This PR solves the problem for the applications that use column metadata to find out whether the column is TIMESTAMP or TIMESTAMP WITH TIME ZONE and based on that info they convert the value they want to store.

vlsi commented:
I guess this changes behaviour, so the only way we can make it is via connection property, so we release a version that adds a property and keeps old behaviour, then keep it for a while, and then release a version that flips the default.

davecramer commented:
@lopata2 can you add a connection property to enable this ?

lopata2 commented:
@davecramer I will try to do that. The same apply also for the TIME and TIME_WITH_ZONE types. Should I do the same change for that data type in this PR or its better to create new PR.

davecramer commented:
> @davecramer I will try to do that. The same apply also for the TIME and TIME_WITH_ZONE types. Should I do the same change for that data type in this PR or its better to create new PR.

You can put them in the same PR.

davecramer commented:
So in order to accept this patch we need to put a configuration parameter to enable it. The default would be to leave the behaviour the same

lopata2 commented:
@davecramer I have added new PGProperty SQL_TYPES_WITH_TIMEZONE that controls the mapping of PG timezone types. The default value is false and the behavior should be the same as it is now. I have also modified the tests.

davecramer commented:
checkstyle is failing...

lopata2 commented:
fixed

vlsi commented on pgjdbc/src/main/java/org/postgresql/PGProperty.java:
```diff
@@ -600,6 +600,16 @@ public enum PGProperty {
+  SQL_TYPES_WITH_TIMEZONE(
+      "sqlTypesWithTimezone",
+      "false",
+      "Enable or disable mapping of PG types with TIMEZONE into SQL types with TIMEZONE."
```
I might be wrong, however, it looks like the description does not explain the meaning of the property.

vlsi commented on pgjdbc/src/main/java/org/postgresql/PGProperty.java:
```diff
@@ -600,6 +600,16 @@ public enum PGProperty {
+   */
+  SQL_TYPES_WITH_TIMEZONE(
+      "sqlTypesWithTimezone",
+      "false",
```
We might consider `enum` style property instead as it might be easier to reason.

davecramer replied:
@vlsi if we used `enum` style what would the two names be ?

davecramer commented:
@vlsi other than the enum style property this is ready for review

JurajBurian commented:
Hello people, we have a problem with this error in our projects. Please try to look at it. Thanks.

davecramer commented:
@vlsi are you still waiting for anything here ?

JurajBurian commented:
Yes, jdbc driver still return wrong metadata of *tz values.

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc2/DatabaseMetaDataTest.java:
```diff
@@ -49,88 +50,106 @@
+  private BinaryMode binaryMode;
+  private boolean sqlTypesWithTimezone;
 
+  public void initDatabaseMetaDataTest(BinaryMode binaryMode, boolean sqlTypesWithTimezone) throws SQLException {
+    final Properties props = new Properties();
+    if (sqlTypesWithTimezone) {
+      PGProperty.USE_WITH_TIMEZONE.set(props, true);
+    }
+    if (binaryMode == BinaryMode.FORCE) {
+      PGProperty.PREPARE_THRESHOLD.set(props, -1);
+    }
+    con = TestUtil.openDB(props);
+
+    this.binaryMode = binaryMode;
+    this.sqlTypesWithTimezone = sqlTypesWithTimezone;
+  }
```
The method has been removed in 4956902c7d3a5217b89d68fafa7fcb6a36e070af
Could you please clarify why do you re-add it back?

davecramer replied:
Ah, ok, my mistake. I should have been more careful rebasing. Let me see the easiest way to undo this.

davecramer replied:
Fixed

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc2/DatabaseMetaDataTest.java:
```diff
@@ -251,12 +274,12 @@ void tables(BinaryMode binaryMode) throws Exception {
-    assertEquals(8, rs.findColumn("TYPE_NAME"));
-    assertEquals(9, rs.findColumn("SELF_REFERENCING_COL_NAME"));
-    assertEquals(10, rs.findColumn("REF_GENERATION"));
+    assertEquals(rs.findColumn("REMARKS"), 5);
+    assertEquals(rs.findColumn("TYPE_CAT"), 6);
+    assertEquals(rs.findColumn("TYPE_SCHEM"), 7);
+    assertEquals(rs.findColumn("TYPE_NAME"), 8);
+    assertEquals(rs.findColumn("SELF_REFERENCING_COL_NAME"), 9);
+    assertEquals(rs.findColumn("REF_GENERATION"), 10);
```
Is this change expected?

vlsi commented on docs/content/documentation/use.md:
```diff
@@ -440,6 +440,9 @@ If we quote them, then we end up sending ""colname"" to the backend instead of "
 Fully qualified class name of the class implementing the AuthenticationPlugin interface. If this is null, the password 
 value in the connection properties will be used.
 
+* **`useWithTimezone (`*boolean*`)`** *Default `false`*\
+Map time with timezone and timestamp with timezone to `TIME_WITH_TIMEZONE` and `TIMESTAMP_WITH_TIMEZONE`
```
I am afraid these naming and the description do not clarify the meaning of the parameter to the end users. It does not clarify the change changes backward compatibility either.

A bit of a problem is that the parameter impacts both `TIME_WITH_TIMEZONE` and `TIMESTAMP_WITH_TIMEZONE`, so it is hard to select a good name.

WDYT of `mapTimestamptzAndTimeTzTypesToJavaSqlWithTimeZone=true|false`?

An alternative option could be `metadata.jdbctype.timetz=TIME | TIME_WITH_TIMEZONE`, `metadata.jdbctype.timestamptz=TIMESTAMP | TIMESTAMP_WITH_TIMEZONE`.



I suggest using `enum` parameter like `mapTimeStampTz

davecramer replied:
I prefer the second option. I'm not a fan of long names.

vlsi commented:
Does the change impact `ParameterMetaData`? If so, I think we need adjust the tests accordingly.

An alternative (additional?) option could be adding the newly added parameters to the test matrix: https://github.com/pgjdbc/pgjdbc/blob/2bfb37db75044d836d57d4165cacadd19c32ae0c/.github/workflows/matrix.js#L128
